src/Analyzers/CSharp/Analyzers/FileBasedPrograms/AppDirectiveHelpers.cs

```diff
@@ -0,0 +1,705 @@
+
+                foreach (var trivia in result.Token.LeadingTrivia)
+                {
+                    ReportErrorFor(trivia);
+                }
+
+                foreach (var trivia in result.Token.TrailingTrivia)
+                {
+                    ReportErrorFor(trivia);
+                }
+            }
+            while (!result.Token.IsKind(SyntaxKind.EndOfFileToken));
+        }
+
+        void ReportErrorFor(SyntaxTrivia trivia)
+        {
+            if (trivia.ContainsDiagnostics && trivia.IsKind(SyntaxKind.IgnoredDirectiveTrivia))
+            {
+                diagnostics.AddError(sourceFile, trivia.Span, CliCommandStrings.CannotConvertDirective);
+            }
+        }
+
+        // The result should be ordered by source location, RemoveDirectivesFromFile depends on that.
+        return builder.ToImmutable();
+    }
+
+    public static void FindDirectives(
+        SourceFile sourceFile,
+        SyntaxTriviaList triviaList,
+        DiagnosticBag diagnostics,
+        ImmutableArray<CSharpDirective>.Builder builder)
+    {
+        // TODO2: netstandard HashSet doesn't have TryGetValue or any other great way of "getting the already present value".
+        var deduplicated = new Dictionary<CSharpDirective.Named, CSharpDirective.Named>(NamedDirectiveComparer.Instance);
+        TextSpan previousWhiteSpaceSpan = default;
+
+        foreach (var (index, trivia) in triviaList.Index())
+        {
+            // Stop when the trivia contains an error (e.g., because it's after #if).
+            if (trivia.ContainsDiagnostics)
+            {
+                break;
+            }
+
+            if (trivia.IsKind(SyntaxKind.WhitespaceTrivia))
+            {
+                Debug.Assert(previousWhiteSpaceSpan.IsEmpty);
+                previousWhiteSpaceSpan = trivia.FullSpan;
+                continue;
+            }
+
+            if (trivia.IsKind(SyntaxKind.ShebangDirectiveTrivia))
+            {
+                var span = GetFullSpan(previousWhiteSpaceSpan, trivia);
+
+                var whiteSpace = GetWhiteSpaceInfo(triviaList, index);
+                var info = new CSharpDirective.ParseInfo
+                {
+                    Span = span,
+                    LeadingWhiteSpace = whiteSpace.Leading,
+                    TrailingWhiteSpace = whiteSpace.Trailing,
+                };
+                builder.Add(new CSharpDirective.Shebang(info));
+            }
+            else if (trivia.IsKind(SyntaxKind.IgnoredDirectiveTrivia))
+            {
+                var span = GetFullSpan(previousWhiteSpaceSpan, trivia);
+
+                var message = trivia.GetStructure() is IgnoredDirectiveTriviaSyntax { Content: { RawKind: (int)SyntaxKind.StringLiteralToken } content }
+                    ? content.Text.AsSpan().Trim()
+                    : "";
+                // TODO: original impl was using more recent span-oriented APIs. Optimal sharing may be tricky.
```

Comment:
Hm, that's sad that we need to target netstandard2.0, but it's probably fine, I don't expect the directive parsing to be perf critical (I just like to use modern APIs).

We also likely should ensure the source package is built against netstandard2.0 inside the sdk repo to verify the compatibility early.